### PR TITLE
Setup for automatic building

### PR DESCRIPTION
**Pull request by SlayterDev**
This PR sets us up to build the configs on merge. This is done using the following process:

When a PR is merged to main:
- Checkout the repo
- Run `index.js`
- Publish the built files to S3
- Commit the newly built files to main

### CODE REVIEW

File: index.js
Comment (jonathanKingston) [inline index.js:9]: Is this needed?

1. Are we envisioning this to still be run by config authors?
2. Should we remove the `generated/` directory from the repo and also add it to the `.gitignore`?

File: .github/workflows/build-and-publish.yml
Comment (jonathanKingston) [inline .github/workflows/build-and-publish.yml:26]: I'm a bit of a beginner at reading these files (we should probably have an additional reviewer that has done more of it like @GioSensation), are we actually calling `node index.js` or `npm start` or similar here? Also if the unit tests just added fail, would this break before uploading to s3?

**Review by jonathanKingston — COMMENTED** [review]
Thanks!
Some questions and the suggestion to get another reviewer IE @GioSensation or similar :).

File: .github/workflows/build-and-publish.yml
Comment (SlayterDev) [reply .github/workflows/build-and-publish.yml:26]: This action only runs _after_ merge so tests should be passing at this stage. Also `line 16` is what runs `index.js` via [this file](https://github.com/duckduckgo/privacy-configuration/blob/main/action.yml)

File: index.js
Comment (SlayterDev) [reply index.js:9]: This is up for debate. When we previously ran the build script in the action we had the issue that a maintainer would:
- generate locally creating one timestamp
- Merge their PR which builds and generates another timestamp

This would cause confusion because the timestamp in the repo was different than the published version.

I'm personally in favor of keeping the generated files in the repo as a quick reference of what's in prod however @kdzwinel also voiced that we could ignore these files in the repo and only build in the actions and just use the online versions for reference (possibly allowing s3 to list the directory their in).

Comment (jonathanKingston) [reply index.js:9]: I think if we want full visibility of published code we could have the CI create a PR to a different repo.

The rationale for removing the generated changes is that it'll reduce the amount of merge conflicts and noise for a reviewer.

Perhaps we can mark the generated files as binaries? Would this reduce the noise in github / git?

File: .github/workflows/build-and-publish.yml
Comment (jonathanKingston) [reply .github/workflows/build-and-publish.yml:26]: I think ideally it should also run tests just in case a broken merge gets put into main?

File: index.js
Comment (kdzwinel) [reply index.js:9]: Uuu… it'd be fancy to release generated files to https://github.com/duckduckgo/privacy-configuration/releases . This action seems to do that: https://github.com/marvinpinto/action-automatic-releases

Comment (SlayterDev) [reply index.js:9]: Wish granted 🪄

File: .github/workflows/build-and-publish.yml
Comment (jonathanKingston) [inline .github/workflows/build-and-publish.yml:29]: Does this get cached? Should we have hour here?

Alternatively we could take it from the generated version in the config?

**Comment by jonathanKingston** [comment]
LGTM
@giosensation or @kdzwinel are you able to double check this as I believe you have some experience with workflows.

File: .github/workflows/build-and-publish.yml
Comment (kdzwinel) [inline .github/workflows/build-and-publish.yml:17]: will whole action fail if tests fail?

Comment (SlayterDev) [reply .github/workflows/build-and-publish.yml:17]: This is my understanding

Comment (kdzwinel) [reply .github/workflows/build-and-publish.yml:17]: we should test that once we merge this, or on some test repo if we want to test it before

Comment (SlayterDev) [reply .github/workflows/build-and-publish.yml:17]: I can't find anything that definitively states this but I think the default behavior of actions is to exit on failure

Comment (GioSensation) [reply .github/workflows/build-and-publish.yml:17]: Yes, actions fail if a script returns something other than 0 or throws.

**Review by GioSensation — APPROVED** [review]
LGTM.

**Review by jonathanKingston — APPROVED** [review]

**Pull request merged**
